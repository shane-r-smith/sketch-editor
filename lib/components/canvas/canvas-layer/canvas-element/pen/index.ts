export * from "./pen";
export * from "./pen.hooks";
