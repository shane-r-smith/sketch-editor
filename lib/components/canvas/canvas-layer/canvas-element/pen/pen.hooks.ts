import type { LineConfig } from "konva/lib/shapes/Line";
import { useMemo } from "react";
import type { CanvasElementProps } from "../canvas-element.types";

export const createPen = ({
  element,
  ...rest
}: CanvasElementProps): LineConfig => ({
  ...rest,
  ...element,
  tension: 0.5,
  lineCap: "round",
  lineJoin: "round",
  globalCompositeOperation: "source-over",
});

export const usePen = (props: CanvasElementProps): LineConfig => {
  return useMemo<LineConfig>(() => {
    return createPen(props);
  }, [props]);
};
