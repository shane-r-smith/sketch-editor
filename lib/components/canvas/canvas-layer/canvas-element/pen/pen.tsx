import { Line } from "react-konva";
import { usePen } from "./pen.hooks";
import type { CanvasElementProps } from "../canvas-element.types";

export function Pen(props: CanvasElementProps) {
  const pen = usePen(props);

  return <Line {...pen} />;
}
