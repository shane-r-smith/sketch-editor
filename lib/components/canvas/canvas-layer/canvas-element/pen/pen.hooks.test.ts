import { describe, it, expect } from "vitest";
import type { CanvasElementProps } from "../canvas-element.types";
import { renderHook } from "@testing-library/react";
import { createPen, usePen } from "./pen.hooks";

// Mock props for testing
const baseProps: CanvasElementProps = {
  element: {
    key: "PEN",
    points: [0, 0, 10, 10],
    stroke: "#000000",
    strokeWidth: 10,
    strokeColourType: "HEX",
  },
  x: 5,
  y: 5,
  id: "pen",
};

describe("createPen", () => {
  it("returns correct LineConfig for pen", () => {
    const result = createPen(baseProps);
    expect(result).toMatchSnapshot();
  });
});

describe("usePen", () => {
  it("returns memoized LineConfig", () => {
    const { result } = renderHook(
      (props: CanvasElementProps) => usePen(props),
      { initialProps: baseProps }
    );
    expect(result.current).toMatchSnapshot();
  });
});
