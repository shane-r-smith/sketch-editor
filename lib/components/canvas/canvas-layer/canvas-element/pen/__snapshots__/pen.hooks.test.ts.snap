// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createPen > returns correct LineConfig for pen 1`] = `
{
  "globalCompositeOperation": "source-over",
  "id": "pen",
  "key": "PEN",
  "lineCap": "round",
  "lineJoin": "round",
  "points": [
    0,
    0,
    10,
    10,
  ],
  "stroke": "#000000",
  "strokeColourType": "HEX",
  "strokeWidth": 10,
  "tension": 0.5,
  "x": 5,
  "y": 5,
}
`;

exports[`usePen > returns memoized LineConfig 1`] = `
{
  "globalCompositeOperation": "source-over",
  "id": "pen",
  "key": "PEN",
  "lineCap": "round",
  "lineJoin": "round",
  "points": [
    0,
    0,
    10,
    10,
  ],
  "stroke": "#000000",
  "strokeColourType": "HEX",
  "strokeWidth": 10,
  "tension": 0.5,
  "x": 5,
  "y": 5,
}
`;
